### PR TITLE
etcd: enable full remote database support

### PR DESCRIPTION
**Pull request by guggero**
Depends on btcsuite/btcwallet#757.

This PR enables full remote database support when using an `etcd` database backend.
This means that everything that used to be in a `bbolt` database now gets written to `etcd`, with an appropriate sub namespace added to the key path.

### CODE REVIEW

File: lncfg/db.go
Comment (joostjager) [inline lncfg/db.go:135]: Ending all namespaces in `db`, maybe it is not necessary as everything in a database backend is data?

File: lncfg/db.go
Comment (joostjager) [inline lncfg/db.go:93]: Why the split between graph and channel? For bbolt that data is stored in the same file and only a single instance was used (somewhere it says `remoteChanDB = localChanDB`). By making it two instances, the global lock semantics change and it would need to be checked that nowhere this logic is relied upon.

Comment (Roasbeef) [reply lncfg/db.go:93]: We did a scan when we did the initial split for the current remote graph portion, and we didn't find any cases where locking between the graph and the channel graph was critical for consistency. Even now we have several databases (watch tower, macaroons, wallet, etc), and if we need to we ensure consistency on restart by applying certain operations again (trying to rebroadcast a transaction as an example), etc. 

Also conceptually, the graph database and the channel database never are linked during routine operations. The only intersection points I can think up off the top are: path finding, and then applying a channel update due to a failed payment. Neither of which need to happen in lock step.

Comment (joostjager) [reply lncfg/db.go:93]: But why the split between these two databases now? We come from bbolt where there was just a single instance serving both data sets. I don't mind, but it seems unnecessary.

Comment (joostjager) [reply lncfg/db.go:93]: On the other hand, splitting them up may reduce lock contention on the global write lock. But still it looks to me like an independent change.

Comment (guggero) [reply lncfg/db.go:93]: I feel like _if_ we're ever going to split them, then _now_ is the time. Adding the namespace to the channel state and wallet DBs will mean that existing `etcd` deployments won't be compatible with `v0.14.0` anyway (which might be another discussion entirely, do we need backward compatibility to `etcd` clusters that were spun up with `v0.13.x`? since it's clearly marked as experimental I hope nobody just migrated their mainnet node?).

Comment (joostjager) [reply lncfg/db.go:93]: If you indeed decide to split now, you will also need to look at the migration logic. Currently that all runs on just the combined database and will need to be split as well?

File: lncfg/db.go
Comment (bhandras) [inline lncfg/db.go]: nit-nit: could be in a previous commit

File: lncfg/db.go
Comment (bhandras) [inline lncfg/db.go]: The nice thing about the `etcd` backend is that we could actually instantiate multiple backends with different namespaces too. This'd also allow higher throughput since all these can work in parallel (each one having it's own commit queue). Not a scope of this PR though, can be done later as an optimization.

Comment (bhandras) [reply lncfg/db.go]: Namespace separation is not required, it's just a nice thing to have and would allow separate snapshotting and backup.

Comment (bhandras) [reply lncfg/db.go]: Crazy idea, we could also separate things into different etcd clusters for even better scalability :)

File: htlcswitch/decayedlog.go
Comment (bhandras) [inline htlcswitch/decayedlog.go:46]: We could move this function under the `kvdb` module to reduce duplication.

File: lncfg/db.go
Comment (bhandras) [inline lncfg/db.go:127]: +1 for creating `bbolt` dbs in a generic function in `kvdb` and then just setting paths here in `lncfg`.

File: watchtower/wtdb/client_db_test.go
Comment (bhandras) [inline watchtower/wtdb/client_db_test.go:834]: Scope creep and could be done in a separate PR but we could also test these with etcd.

File: kvdb/etcd/config.go
Comment (bhandras) [inline kvdb/etcd/config.go:35]: Awesome! Please disregard my previous comment about this.

File: lncfg/db.go
Comment (bhandras) [inline lncfg/db.go:142]: Great, looks like my other comment is also implemented :) More scalability!

**Review by bhandras — COMMENTED** [review]
Awesome work @guggero !! My main comment is whether we should also add interfaces by function in this PR and if we should also support testing of all these DBs with all backends. Other than that, LGTM!

**Comment by joostjager** [comment]
I noticed one thing while running the postgres `stateless_init` itest. In this test, the password is changed. But during the handling of this call, the macaroon service and the macaroon db connection is closed in https://github.com/lightningnetwork/lnd/blob/c733c139e95a6ef4e5f9ac88b43328ac96c333ef/walletunlocker/service.go#L653.

This leads to a failure for the subsequent rpc call when validating the macaroon.

File: chainreg/chainregistry.go
Comment (guggero) [reply chainreg/chainregistry.go:73]: I'm not sure if this is exactly what you had in mind. But I added a refactor commit (c95fe875) to separate the height hint DB more (and not be dependent on the `channeldb.DB` struct.

File: lncfg/db.go
Comment (guggero) [reply lncfg/db.go]: Yes, the namespacing is in the last commit to make the changes very contained. This will allow us to separate things out even further in the future, correct. Maybe not only different clusters but also different backend technologies? E.g. something fast for the graph and something more on the reliable side for the channel state?

File: htlcswitch/decayedlog.go
Comment (guggero) [reply htlcswitch/decayedlog.go:46]: My main concern was to not create more package dependencies between `lncfg` and `macaroons`/`watchtower`. But since we move all other DB related stuff to `lncfg` I decided to also move the database filename constants there to avoid those dependencies.

File: lncfg/db.go
Comment (guggero) [reply lncfg/db.go:127]: Went with something slightly different, let me know what you think.

File: watchtower/wtdb/client_db_test.go
Comment (guggero) [reply watchtower/wtdb/client_db_test.go:834]: I'll keep that in mind for a follow-up PR :+1:

File: lncfg/db.go
Comment (guggero) [reply lncfg/db.go]: This is intentional to make it more obvious this is a namespacing thing (for anyone just looking at the key names). I'd prefer to leave it that way, even if it sounds superfluous.

File: lncfg/db.go
Comment (guggero) [reply lncfg/db.go:93]: The schema migration is definitely a good argument. I decided to not do the split in this PR. But I'll start a new PR on top of this one where I look into how much work would need to go into this.

**Comment by guggero** [comment]
> I noticed one thing while running the postgres stateless_init itest. In this test, the password is changed. 

I addressed this in the version I pushed last week. The macaroon service no longer closes the DB backend now (since it is also not responsible for creating it in the first place). If you still see this with the current version, please let me know.

File: lncfg/db.go
Comment (joostjager) [inline lncfg/db.go:125]: I couldn't find any use of this bool anymore besides modifying a log line, so perhaps it can be removed now? Also replicated isn't accurate, because it can be a single instance remote db backend.

File: lnd.go
Comment (joostjager) [inline lnd.go:1648]: Isn't this applying migrations twice to the same underlying db now?

File: lnd.go
Comment (joostjager) [inline lnd.go:1666]: Define constants?

File: lncfg/db.go
Comment (bhandras) [reply lncfg/db.go]: Exactly

**Review by bhandras — APPROVED** [review]
Changes LGTM barring Joost's comment on the migration. Great work Oliver!

**Comment by bhandras** [comment]
> Changes LGTM barring Joost's comment on the migration. Great work Oliver!

There's still something off with the itests it seems. We may need another round later on.

File: lncfg/db.go
Comment (guggero) [reply lncfg/db.go:125]: The decision for the wallet is removed in the second to last commit. But I think the log statement is nice to have. I renamed it to `Remote` though to be more accurate. And I removed it from the root package's `databaseInstances` struct where it's definitely not needed anymore.

File: lnd.go
Comment (guggero) [reply lnd.go:1648]: Hmm, you're right. I think we can't split it here yet until we do the actual separation of the two databases on the code level. Also added a TODO.

File: lnd.go
Comment (joostjager) [inline lnd.go:1644]: If this is the only place where the non-decorated  closes are called, isn't it more clear to just write out the non-decorated closes here and skip the close func map?

**Review by joostjager — APPROVED** [review]
Looks good. Only comment is that the commit order is such that not every commit seems to be safe to run. For example the etcd namespacing is introduced after all data is already stored in the same etcd database in a previous commit.

File: lncfg/db.go
Comment (joostjager) [inline lncfg/db.go:260]: What about closing `boltBackend` in this case?

Comment (guggero) [reply lncfg/db.go:279]: Ah yes, good point. Added a cleanup list/function here as well.

**Comment by guggero** [comment]
PR now depends on btcsuite/btcwallet#757, had to fix some issues around DB transactions in the wallet. But all integration tests should now work (if not affected by known flakes).

File: channeldb/db.go
Comment (bhandras) [inline channeldb/db.go:229]: nit: Could be squashed with the previous commit.

File: lnd.go
Comment (joostjager) [inline lnd.go:1608]: Don't you want to return the close funcs from `GetBackends` to not repeat yourself?

File: routing/graph.go
Comment (joostjager) [inline routing/graph.go:32]: >To make sure we don't interfere with the kvdb logic when it comes to remote backends such as etcd, we use the kvdb.View method instead of keeping a long-lived read TX open.

What kind of interference is this exactly, a dead lock?

**Comment by joostjager** [comment]
@guggero is this PR now blocked by #5547 ?

**Comment by guggero** [comment]
> @guggero is this PR now blocked by #5547 ?

Yes, and the btcwallet one.

File: channeldb/db.go
Comment (guggero) [reply channeldb/db.go:229]: I wanted to give this special attention in case it has any side effects I wasn't aware of, that's why it is in its own commit.

File: routing/graph.go
Comment (guggero) [reply routing/graph.go:63]: Okay, I think I drop the commit for the moment since we have identified the actual culprit that breaks the itests. We'll need to address the path finding later on anyway.

File: routing/graph.go
Comment (guggero) [reply routing/graph.go:32]: Okay, I think I drop the commit for the moment since we have identified the actual culprit that breaks the itests. We'll need to address the path finding later on anyway.

File: macaroons/store.go
Comment (joostjager) [inline macaroons/store.go:81]: What I ran into is that this line 

https://github.com/lightningnetwork/lnd/blob/fb17bf2bb5cd49c01501e1842ed8082df75d5877/macaroons/store.go#L100

should not pass in `r`, but `r.Backend`. Otherwise the extended backend isn't recognized. Best to panic in `BeginReadTx` to always catch these things.

File: lnd.go
Comment (guggero) [reply lnd.go]: Yeah, makes a lot of sense. Was able to simplify this quite a bit. What do you think?

File: macaroons/store.go
Comment (guggero) [reply macaroons/store.go:81]: Ah yes, makes sense. Good catch!

You mean to add a panic to `BeginReadTx` of the etcd and postgres implementations because there we always expect an extended DB? Or where exactly would you put that panic?

**Comment by guggero** [comment]
Rebased after merging #5547. Still depends on btcsuite/btcwallet#757 but now has the "single writer" flag turned on for the wallet DB instance.

File: macaroons/store.go
Comment (joostjager) [reply macaroons/store.go:81]: Yes, I put it there indeed.

**Review by joostjager — APPROVED** [review]
Changes look good. I didn't look too closely at or tested the watchtower functionality.

What I did do is run the etcd bottle benchmark on this branch, no errors so far.

File: lncfg/db.go
Comment (joostjager) [inline lncfg/db.go:241]: I still feel pretty nervous about just flipping all of the databases besides this one to concurrent writing. Itests may pass, but in general those don't really test concurrent scenarios. How certain is it that those would catch concurrency issues?

Maybe it is worth considering to start out more conservative here. Things like macaroondb and the towerdbs - do those really need concurrent writing from the start?

As mentioned elsewhere, I'd prefer to not do any concurrent writing at all for the first etcd release.

File: macaroons/store.go
Comment (joostjager) [reply macaroons/store.go:81]: Also with the dropped routing tx commit, there is still a BeginReadTx in `newDbRoutingTx`

File: lnd.go
Comment (bhandras) [inline lnd.go]: nit: calling the `Close` multiple times may result in error logs.

File: channeldb/db.go
Comment (bhandras) [reply channeldb/db.go:229]: SGTM

File: lntest/itest/log_error_whitelist.txt
Comment (bhandras) [inline lntest/itest/log_error_whitelist.txt:143]: nit: I think we can filter out the context cancelled error messages when we call the close funcs.

**Review by bhandras — APPROVED** [review]
LGTM (barring the btcwallet PR), awesome work @guggero! Great progress towards making LND more reliable! :1st_place_medal:

**Comment by joostjager** [comment]
>What I did do is run the etcd bottle benchmark on this branch, no errors so far.

The benchmark crashed after 15k payments. I used https://github.com/bottlepay/lightning-benchmark/tree/etcd-updates

`rpc error: code = Unavailable desc = etcdserver: request timed out`

File: lncfg/db.go
Comment (joostjager) [inline lncfg/db.go:190]: I also ran into this line: 

https://github.com/lightningnetwork/lnd/blob/f4da59ec1cd960650384578101084ad6557f585d/batch/batch.go#L74

Here the error isn't checked, because `failIdx >= 0` must mean that an error occurred. Only with the new tx retries, this is no longer true? If a serialization error occurred somewhere and the retry succeeded, `failIdx` is still set and the error logic is applied.

Comment (joostjager) [reply lncfg/db.go:190]: Has etcd been tested with and without single writer on for example the benchmark to see how much the performance gain actually is? Serializable isolation is not necessarily faster if a lot of collisions happen.

File: macaroons/store.go
Comment (guggero) [reply macaroons/store.go:81]: I went ahead and implemented the idea suggested by @bhandras to add the `View()` and `Update()` methods directly to the `walletdb.DB` interface. That way we don't do any runtime interface checks anymore which removes the need for a panic.

A manual `BeginReadTx` is okay I think since that cannot cause another TX to fail as it doesn't change any data.

File: lncfg/db.go
Comment (guggero) [reply lncfg/db.go:241]: Discussed offline. We'll add more bug fixes and tests on top of this PR to make sure everything is concurrency safe. Also having multiple weeks of Travis runs with the lock removed should help us find at least some of those hidden bugs (that we want to fix anyway in the mid term).

File: lnd.go
Comment (guggero) [reply lnd.go]: Oops, this was a left over rebase artifact. Each backend's `Close()` function should now only be called once.

File: lncfg/db.go
Comment (guggero) [reply lncfg/db.go:190]: Good find! We'll run extensive benchmark and other tests once this PR and some of the other fixes have landed in master.
I've started a new general issue to keep track of all the things that will be found in the process: https://github.com/lightningnetwork/lnd/issues/5557

File: channeldb/db.go
Comment (guggero) [reply channeldb/db.go:229]: Was able to remove this commit after the latest set of updates (following your idea).

File: macaroons/store.go
Comment (joostjager) [reply macaroons/store.go:81]: I remember reading about serializable isolation and iirc there was also something with a read failing under certain conditions.

File: lntest/itest/test_harness.go
Comment (bhandras) [inline lntest/itest/test_harness.go:244]: nit:`mineBlocksSlow`

**Review by bhandras — APPROVED** [review]
LGTM, let's get this merged!

File: macaroons/store.go
Comment (joostjager) [reply macaroons/store.go:81]: @guggero are you sure that read-only txes cannot fail? I'd think that a db engine can overlap txes in such a way that a read-only has to give way to a write.

Comment (joostjager) [reply macaroons/store.go:81]: I believe this is where I saw it: https://wiki.postgresql.org/wiki/SSI

Last example, "rollover"

Comment (guggero) [reply macaroons/store.go:81]: Hmm, good point. We are going to address the issue of manual `BeginReadTx` in https://github.com/lightningnetwork/lnd/pull/5595 though.

**Pull request merged**
